Product Collection: Add "Related by" setting to Related Products collection

imanish003 opened the pull request:
<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #52021

This PR adds "Related by" inspector control to Related Products collection. This setting allows to select which criteria will be used to find related products. There are two checkboxes:
- Categories
- Tags

![image](https://github.com/user-attachments/assets/391dbda6-3ed2-48f5-8251-1560f3889e75)


By default, both "Categories" and "Tags" options are checked. This means that related products will be shown based on both criteria. Users can uncheck any option to show related products based on their preferred criteria only.

For example:
- If only "Categories" option is checked, related products will be shown based on categories only.
- If only "Tags" option is checked, related products will be shown based on tags only.
- If both options are checked, related products will be shown based on both categories and tags.

Note: "Attributes" option is not included in this iteration as it requires significant development effort and the value it provides is not as high compared to Categories and Tags.


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

**Prerequisites:**
- For testing purposes, you will need products which are related by Categories and Tags.
- To test "Tags" option, you will need products which are related by Tags. For example:
	- Create a new tag names `tag_a`.
	- Assign this tag to a product (example: "Album") for which we wanna see related products.
	- Now assign the same tag to other products (example: "Polo", "Sunglasses") which should be shown as related products for "Album" product.
- Similarly, to test "Categories" option, you will need products which are related by Categories. For example:
	- Create a new category names `cat_a`.
	- Assign this category to a product (example: "Album") for which we wanna see related products.
	- Now assign the same category to other products (example: "Belt", "Cap") which should be shown as related products for "Album" product.

**Scenario 1: Test "Related by" setting with Tags and Categories**

1. Create a new post and add Product Collection block.
2. Select "Related Products" collection.
3. Select a product (example: "Album") - you will need to select a specific product to see related products.
4. Verify that by default both "Categories" and "Tags" options are checked.
5. Uncheck "Categories" option.
6. Verify that related products are shown based on "Tags" option only on Editor side. (example: "Polo", "Sunglasses")
	- Publish the post and verify same on the frontend.
7. Now only check "Categories" option and verify that related products are shown based on "Categories" option only on Editor side. 
	- Publish the post and verify same on the frontend.
8. Now check both "Categories" and "Tags" options and verify that related products are shown based on both "Categories" and "Tags" options on Editor side.
	- Publish the post and verify same on the frontend.

**Scenario 2: Verify tracking for "Related by" control**

1. Preparations
	1. Create a new post
	2. Open your browser Dev Tools (press Cmd + Opt + I on Mac)
	3. In the Console tab, paste and run this code: `localStorage.setItem( 'debug', 'wc-admin:*' )`
	4. In the Console settings, enable "Verbose" logging level to see detailed tracking events
2. In the post, Insert Product Collection block and choose "Related Products" collection.
3. Interact with "Related by" control in Inspector and verify that the following tracking event appears in the Console tab:
   ```
   wc-admin:tracks recordevent wcadmin_blocks_product_collection_inspector_control_clicked {
     collection: 'woocommerce/product-collection/related',
     location: 'post', 
     filter: 'related-by'
   }
   ```
4. Also verify that when you change "Categories" or "Tags" options, the following tracking event appears in the Console tab:
   ```
   wc-admin:tracks recordevent blocks_product_collection_inspector_control_related_by_changed {
     categories: true/false,
     tags: true/false
   }
   ```
<img width="1311" alt="image" src="https://github.com/user-attachments/assets/dce9e404-47e5-479f-987c-7e247b0f4473">


<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [x] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Product Collection: Add "Related by" setting to Related Products collection
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

imanish003 committed:
feat: Add related by setting to related products collection

This commit adds a new setting to the related products collection that allows
users to control how products are related. Users can choose to relate products
by categories, tags, or attributes.

imanish003 committed:
Remove attributes from related by control

This commit removes the attributes option from the related by control in the
Product Collection block. This option was not being used in the backend and
was potentially misleading to users.

Changes made:
- Removed attributes checkbox from RelatedByControl component
- Removed attributes from RelatedBy type definition
- Cleaned up PHP code by using PHP_INT_MAX instead of magic number

imanish003 committed:
feat: Set default values for related by setting

This commit sets the default values for the relatedBy setting in the Product
Collection block. By default, products will be related by both categories and
tags.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks, woocommerce

imanish003 commented:
Hi Folks,

I've marked this PR as ready for review. The implementation of the "Related by" setting for Related Products collection is complete and working as expected. I'll be adding tests while the PR is under review to ensure everything functions correctly.

Looking forward to your feedback!

imanish003 committed:
Add tracking for related-by filter

imanish003 committed:
Add e2e test for "Related by" setting in Related Products collection

imanish003 committed:
Merge branch 'trunk' into add/52021-product-collection-add-related-by-setting-to-related-products-collection

imanish003 commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/constants.ts:
```diff
@@ -64,6 +64,10 @@ export const DEFAULT_QUERY: ProductCollectionQuery = {
 	timeFrame: undefined,
 	priceRange: undefined,
 	filterable: false,
+	relatedBy: {
+		categories: true,
+		tags: true,
+	},
```
As discussed [here](https://a8c.slack.com/archives/C05UMTA7W0P/p1730808930132549?thread_ts=1730807011.817329&cid=C05UMTA7W0P), I have decided to not include "Attributes" option in this iteration as it requires good amount of efforts and value it provides is not that high comparing to Categories and Tags.

imanish003 commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php:
```diff
@@ -1980,11 +1980,26 @@ function ( $collection_args ) {
 					100
 				);
+
+				remove_filter( 'woocommerce_product_related_posts_relate_by_category', $category_callback, PHP_INT_MAX );
+				remove_filter( 'woocommerce_product_related_posts_relate_by_tag', $tag_callback, PHP_INT_MAX );
```
I'm using WooCommerce's filter hooks to control which taxonomies (categories/tags) are used for finding related products, then removing the filters after getting results. I chose this over creating custom queries to leverage WooCommerce's existing optimized logic.

Does this approach make sense? Any concerns about using filters this way?

octaedro commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/related-by-control.tsx:
```diff
@@ -0,0 +1,59 @@
+import {
+	CheckboxControl,
+	PanelBody,
+	// @ts-expect-error missing types.
```
Are we using VStack in this file?

octaedro commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/related-by-control.tsx:
```diff
@@ -0,0 +1,59 @@
+import {
+	CheckboxControl,
+	PanelBody,
+	// @ts-expect-error missing types.
+	__experimentalVStack as VStack,
```
Are we using VStack in this file?

imanish003 replied:
Thanks for catching that! I initially used VStack but later simplified it to a regular `div` element since I didn't need any of VStack's specific layout features. I will remove it.

octaedro commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/related-by-control.tsx:
```diff
@@ -0,0 +1,59 @@
+const RelatedByControl = ( {
+	query,
+	setQueryAttribute,
+	trackInteraction,
```
Currently, we are only recording the props `collection`, `filter` and `location` under the event `blocks_product_collection_inspector_control_clicked`. However, I believe it would be beneficial to also record whether `Categories` or `Tags` are selected. Perhaps we can track this under a separate event, such as `blocks_product_collection_inspector_control_related_selected` or something similar.

imanish003 replied:
I was also thinking about this but wasn't sure if we should track this separately. Let's track this under a separate event. I have added the changes in 59a19d7.

octaedro commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/related-by-control.tsx:
```diff
@@ -0,0 +1,59 @@
+	setQueryAttribute,
+	trackInteraction,
+}: QueryControlProps ) => {
+	const relatedBy = query.relatedBy as RelatedBy;
```
Completely nitpick, but since this is an [optional prop](https://github.com/woocommerce/woocommerce/pull/52580/files#diff-2730289cd10e5059fa0c04f43f2d0c358137074d357d898eccdfbb7fd0e3d890R120) it would make sense to add an optional chaining operator here (`query?.relatedBy`).

octaedro reviewed (commented):
Great job @imanish003! This is testing well on my end, and the code looks good. I just left a couple of minor suggestions.

imanish003 committed:
Remove unused import for VStack

imanish003 committed:
Add optional chaining operator for query.relatedBy

Since relatedBy is an optional prop, add the optional chaining operator to
safely access it and prevent potential undefined property access errors.

imanish003 committed:
Add tracking for related by changes in Product Collection block

This commit adds tracking for when users change the "Related by" settings in
the Product Collection block inspector controls. It records whether categories
and tags are enabled or disabled.

imanish003 committed:
Merge branch 'trunk' into add/52021-product-collection-add-related-by-setting-to-related-products-collection

imanish003 commented:
Hey @octaedro, Thanks for your detailed feedback! I have made the changes and it's ready for another review 🙇‍♂️

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/related-by-control.tsx:
```diff
@@ -0,0 +1,67 @@
+	};
+
+	return (
+		<PanelBody title={ __( 'Relate by', 'woocommerce' ) }>
```
I saw designs have "Relate by" but I guess it's a typo:

```suggestion
		<PanelBody title={ __( 'Related by', 'woocommerce' ) }>
```

And if not, code reference should reflect that and we should change `relatedBy` to `relateBy` 😅

imanish003 replied:
Good catch! As discussed on Slack, I have made the changes in [3cd67d1](https://github.com/woocommerce/woocommerce/pull/52580/commits/3cd67d1e019116f77b84cfb662a45e3c68f0eacb) 😀

kmanijak commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php:
```diff
@@ -1980,11 +1980,26 @@ function ( $collection_args ) {
+				add_filter( 'woocommerce_product_related_posts_relate_by_tag', $tag_callback, PHP_INT_MAX );
+
 				$related_products = wc_get_related_products(
 					$collection_args['relatedProductReference'],
```
I know it's not in scope of this PR but would be good to keep collection name consistent across code:

```suggestion
					$collection_args['relatedProductsReference'],
```

That would be applied in multiple places.

imanish003 replied:
AFAIU This name is directly related to the `productReference` attribute name. Therefore, it's called `relatedProductReference`. But Happy to change the name if you think it's better to name it `relatedProductsReference`. I don't have strong feelings either way 😀

kmanijak replied:
Ah I see. 😅 Nvm!

imanish003 committed:
Update panel title from "Relate by" to "Related by"

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/related-by-control.tsx:
```diff
@@ -0,0 +1,67 @@
+
+		trackInteraction( CoreFilterNames.RELATED_BY );
+
+		recordEvent(
+			'blocks_product_collection_inspector_control_related_by_changed',
+			{
+				categories: newRelatedBy?.categories || false,
+				tags: newRelatedBy?.tags || false,
+			}
+		);
```
Do we want a new event just for this? 🤔 If needed maybe we can collect more data within the interaction above although the assumption was we don't need information what's the value merchants choose but the number of interaction matters.

imanish003 replied:
While I don't have a strong opinion on this, I see potential value in keeping the tracking data:

1. It could help us understand how merchants are using the related products feature, particularly which relationship types (categories vs tags) they prefer

2. Even if we don't have immediate plans to analyze this data, having it available could be useful for future decision-making around feature improvements

3. Since we already have the tracking infrastructure in place, the overhead of collecting this additional data point is minimal

That said, I'm open to removing it if others feel it's unnecessary. I mainly added it based on @octaedro's [suggestion](https://github.com/woocommerce/woocommerce/pull/52580#discussion_r1831126066).

kmanijak replied:
Personally I'm not convinced 😅 I had discussion about the tracking with @octaedro and we clearly have different approach 😅 

> It could help us understand how merchants are using the related products feature, particularly which relationship types (categories vs tags) they prefer

I can't think of single usage atm. Because let's imaging scenario user plays with the setting:
- clicks categories 8 times
- clicks tags 2 times
They end up with default setup. We cannot conclude from those numbers categories are more popular.

To have a meaningful data what merchants prefer, we should actually extend logic by @xristos3490 from this PR: https://github.com/woocommerce/woocommerce/pull/46466.

> Even if we don't have immediate plans to analyze this data, having it available could be useful for future decision-making around feature improvements

I see, can't argue with that. Although I prefer more purposeful approach where we say "We want to know X, let's measure Y". In here, I cannot see potential use cases is all 😅 

> Since we already have the tracking infrastructure in place, the overhead of collecting this additional data point is minimal

True implementation-wise, but we need to remember collecting data long-term comes with the cost (storing etc.)

---

Overall I won't be fighting for removal of this however:
- I'd love to understand one potential use case for it
- I believe extending https://github.com/woocommerce/woocommerce/pull/46466 instead would give us more value.

octaedro replied:
Hi @imanish003 and @kmanijak, this is an interesting thread.

Here are my two cents. Collecting data for new functionalities is usually a good idea. In this case, depending on how users interact with the checkboxes, we can prioritize or deprioritize the work done there.

If nobody interacts with the checkboxes, we'll know that improving them is not a priority. Conversely, if we see a high level of interaction, it may indicate that the component is not as intuitive as we thought. In such cases, adding a text explanation on how it works could be beneficial.

> True implementation-wise, but we need to remember collecting data long-term comes with the cost (storing etc.)

While it is true that any data collection process involves costs associated with storage, modern technologies and efficient data management practices make these costs manageable and often offset by the value that the data provided.

Having said that, if you think that collecting this data is not useful, I'm okay with not recording it or only recording the interactions, regardless of the selected items.

I would like to know @j111q's opinion on this.

kmanijak replied:
Don't get me wrong! I'm ok with collecting data but I'm suggesting different approach (tracking outcome) which IMO gives more information and is in sync with how we track all the other filters/controls.

imanish003 replied:
Great to see such a productive discussion. Everyone is raising their concerns and providing thoughtful perspectives on the tracking approach.  Thank you 🙌🏻

> To have a meaningful data what merchants prefer, we should actually extend logic by @xristos3490 from this PR: #46466.

I agree that the current tracking approach doesn't provide meaningful insights into merchant behavior. The saved filter values in post would be much more valuable for understanding how the feature is actually being used.

After reviewing the PR by @xristos3490, I notice it only tracks which features/filters are enabled, without capturing their configured values. While extending the tracking to include these values would be valuable, I believe this enhancement deserves its own dedicated task rather than being added to this PR. Because we should maintain consistency by tracking filter values across all Product Collection controls, not just this one. Also, It would significantly expand this PR's scope beyond its core purpose.

I suggest we create a separate task to implement comprehensive filter value tracking. What are your thoughts on this approach? CC: @kmanijak @octaedro

> If nobody interacts with the checkboxes, we'll know that improving them is not a priority.

@octaedro We actually already have tracking in place for this via the `blocks_product_collection_inspector_control_clicked` event [here](https://github.com/woocommerce/woocommerce/blob/3cd67d1e019116f77b84cfb662a45e3c68f0eacb/plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/related-by-control.tsx#L33). This existing tracking should give us good visibility into whether users are interacting with these controls or not. Isn't it?

octaedro replied:
> Don't get me wrong! I'm ok with collecting data but I'm suggesting different approach (tracking outcome) which IMO gives more information and is in sync with how we track all the other filters/controls.

Yes, that approach aligns well with our goals and seems worth implementing.

> I suggest we create a separate task to implement comprehensive filter value tracking. What are your thoughts on this approach?

Good idea! It would provide very valuable data.

> We actually already have tracking in place for this via the blocks_product_collection_inspector_control_clicked event [here](https://github.com/woocommerce/woocommerce/blob/3cd67d1e019116f77b84cfb662a45e3c68f0eacb/plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/related-by-control.tsx#L33). This existing tracking should give us good visibility into whether users are interacting with these controls or not. Isn't it?

Agreed, that makes sense. Recording the `blocks_product_collection_inspector_control_clicked` event along with the additional tracking method suggested above should be enough.

imanish003 replied:
Based on above discussion, I have created a new issue #52696 to implement control values tracking.

kmanijak reviewed (commented):
I tested the PR and it's working just fine! I left couple of nitpicks/questions 🙌

imanish003 committed:
Remove `blocks_product_collection_inspector_control_related_by_changed` track event

imanish003 commented:
Hey @kmanijak @octaedro,

As discussed, I have created a new issue #52696 to implement values tracking and removed `blocks_product_collection_inspector_control_related_by_changed` event from this PR. Can you please review this PR again? It would be great to merge it today 😀

octaedro reviewed (approved):
Thank you @imanish003 for addressing the changes!
This is testing well on my end, and the code looks good to me 🚀 
Good job!

octaedro commented:
Hey @imanish003,
I'm going to merge this since the code freeze is in a few hours.
